### ISSUE TITLE
Correctly joining `LD_LIBRARY_PATH` and similar bash variables throughout nixpkgs

### ISSUE THREAD

**Issue by nh2** [NixOS/nixpkgs#144646]
There are places in nixpkgs where the pattern `LD_LIBRARY_PATH=somedir:$LD_LIBRARY_PATH` is used.

That is a security risk (see  [article](https://jdhao.github.io/2021/07/03/ld_library_path_empty_item/)).

This issue is about removing all such bad patterns from nixpkgs, for `LD_LIBRARY_PATH`, `PATH`, and potentially other variables.

###### Security risk explanation

If `$LD_LIBRARY_PATH` hasn't been set before (is undefined or the empty string), then the variable expansion will end in `...:`, thus having an empty entry on the right side.

Having an empty entry in it puts the _current directory_ onto `LD_LIBRARY_PATH` (e.g.  `somedir:` is the same as `somedir:.`) , thus loading `.so` files and executing code that the user may not wish to execute.

For example, loading downloaded `.so` files from the Internet in a directory the user `cd`d into.

###### Nixpkgs example

Here is one such case that even wraps a packaged program permanently with such a wrong `LD_LIBRARY_PATH`:

https://github.com/NixOS/nixpkgs/blob/164173d0fc914ec5f70b67cf1c8b2f03142cb7fc/pkgs/applications/office/jabref/default.nix#L53-L55

## Correct variable concatenation in plain shell

There are 2 standard **correct patterns** that prepend the separator `:` only if `LD_LIBRARY_PATH`, or any variable `V`, is not itself empty:

1. Conditionally expanding `:$V` (the combination of `:` and `$V`), from https://unix.stackexchange.com/questions/162891/append-to-path-like-variable-without-creating-leading-colon-if-unset:
   ```sh
   export V="somethingPrepended${V:+:${V}}"
   export LD_LIBRARY_PATH="somedir${LD_LIBRARY_PATH:+:${LD_LIBRARY_PATH}}"
   ```
2. Conditionally expanding just the `:`, as currently more common in nixpkgs ([example for appending here](https://github.com/NixOS/nixpkgs/blob/164173d0fc914ec5f70b67cf1c8b2f03142cb7fc/pkgs/development/libraries/silgraphite/graphite2.nix#L39)):
   ```sh
   export V="somethingPrepended${V:+:}${V}"
   export LD_LIBRARY_PATH="somedir${LD_LIBRARY_PATH:+:}$LD_LIBRARY_PATH"
   ```

I personally think that both are quite unreadable, and easy to get wrong.

## Correct variable concatenation in nix bash scripts

Further, in nix multi-line string literals (`''`), escaping of `$` is needed, making it `''$`:

```nix
''
   export V="somethingPrepended''${V:+:''${V}}"
   export LD_LIBRARY_PATH="somedir''${LD_LIBRARY_PATH:+:''${LD_LIBRARY_PATH}}"

   export V="somethingPrepended''${V:+:}''${V}"
   export LD_LIBRARY_PATH="somedir''${LD_LIBRARY_PATH:+:}''$LD_LIBRARY_PATH"
''
```

The needed escaping makes it even more difficult to get it right, which probably contributes to this correct way not being used everywhere.

## What could be done to improve this?

One or multiple of the below:

* Provide a nix function that makes this easier (I suggest one below), and enforce it in reviews.
* Provide a shell function that makes this easier (similar to `substituteInPlace`).
* Enforce one of the above safe patterns in reviews (may be tough, as they are difficult to remember).
* Add a linter that can find such issues in reviews automatically.

**Comment by nh2** [comment]
For my own usage, I've written this nix function as one possible implementation of the idea

> Provide a nix function that makes this easier (I suggest one below), and enforce it in reviews.

```nix
# Returns a bash string expression concat two values with a separator,
# skipping the separator if any of the values is empty or undefined.
# Based on https://unix.stackexchange.com/questions/162891/append-to-path-like-variable-without-creating-leading-colon-if-unset
# and then extended to both values.
#
# This is expecially useful for variables like LD_LIBRARY_PATH,
# where having a `stuff:` entry with one side of the `:` being empty
# will put the current directory onto that search path, making
# the running of commands extremely slow from e.g. networked dirs.
#
# The values are put into double-quotes (`""`) so they may contain
# shell variables that will be evaluated.
# The returned string is double-quoted.
#
# Example usage in a script:
#     ''
#       export LD_LIBRARY_PATH=${bashConcatWithSeparator "/my/path" ":" "$LD_LIBRARY_PATH"}
#     ''
bashConcatWithSeparator = left: separator: right:
  # Idea:
  #     1. Prepend each side with the separator.
  #     2. Join them.
  #     3. chop off separator's length from the front.
  #
  # Example:
  #     SEP=':'; JOINED="${A:+$SEP$A}${B:+$SEP$B}"; echo "${JOINED:${#SEP}}"
  #
  # Where:
  # * `${A:+$SEP$A}` returns ":$A" if `A` isn't empty, and empty string otherwise
  # * `${#SEP}` returns the length of `SEP`, e.g. `3`
  # * `${JOINED:3}` returns `JOINED` with 3 chars chopped off its front
  #
  # Exhaustive example:
  #     $ A=;  B=;  SEP=':'; JOINED="${A:+$SEP$A}${B:+$SEP$B}"; echo "${JOINED:${#SEP}}"
  #
  #     $ A=a; B=b; SEP=':'; JOINED="${A:+$SEP$A}${B:+$SEP$B}"; echo "${JOINED:${#SEP}}"
  #     a:b
  #     $ A=;  B=b; SEP=':'; JOINED="${A:+$SEP$A}${B:+$SEP$B}"; echo "${JOINED:${#SEP}}"
  #     b
  #     $ A=a; B=;  SEP=':'; JOINED="${A:+$SEP$A}${B:+$SEP$B}"; echo "${JOINED:${#SEP}}"
  #     a
  # Generate the string (including the returned quotes) like e.g.:
  #     "$(A="a"; B="b"; SEP=":"; JOINED="${A:+$SEP$A}${B:+$SEP$B}"; echo "${JOINED:${#SEP}}")"
  ''"$(A="${left}"; B="${right}"; SEP="${separator}"; JOINED="''${A:+$SEP$A}''${B:+$SEP$B}"; echo "''${JOINED:''${#SEP}}")"'';
```

As mentioned above, it is used like this:

```nix
''
  export LD_LIBRARY_PATH=${bashConcatWithSeparator "somedir" ":" "$LD_LIBRARY_PATH"}
''
```

and works for prepending (as above), and for appending:

```nix
''
  export LD_LIBRARY_PATH=${bashConcatWithSeparator "$LD_LIBRARY_PATH" ":" "somedir"}
''
```

I'd like some feedback on this function specifically (correctness review), and whether others could imagine to use this, or would prefer other solutions.

**Comment by r-burns** [comment]
IMO stdenv's addToSearchPath (when available) is the most correct way to do this.

**Comment by nh2** [comment]
> IMO stdenv's addToSearchPath (when available) is the most correct way to do this.

For non-package use cases (like NixOS modules, e.g. `systemd.services` definitions), that counts as "not available", right? Or could it still be used there?

**Comment by r-burns** [comment]
Yes, in cases like that it would be useful to have your helper function to handle the variable expansion wizardry. Or alternatively, we could hookify addToSearchPath et al. so it's possible to `source` in non-stdenv scenarios.

**Comment by tobiasBora** [comment]
Do you think it would be possible to detect automatically these bad usage and display a Warning to the user?

**Comment by andersk** [comment]
I did a treewide pass for this last year at #76804, then fixed a new instance at #89243. It would be nice to have an automatic CI check for this.

**Comment by andersk** [comment]
@tobiasBora There’s a proposal to patch glibc at #76802, but nobody’s written the patch yet.
